File: clients/macos/Packages/WindexKit/Sources/WindexKit/Models/Param.swift
import Foundation

/// One typed, bounded, self-describing parameter — the shape every form renders.
///
/// This is the Swift mirror of `windex.schema.param.Param.describe()`
/// (`src/windex/schema/param.py`). It is decoded **generically**: nothing here
/// names a windex setting. That is the whole point — the same struct backs
/// editable settings, Run arguments, and Pipeline Module configuration, so
/// `SchemaForm` is written once against `Param` and the Pipeline composer drops
/// straight onto it without a second value renderer.
///
/// Two server rules a client must not paper over:
///
/// * **Clamp, don't reject.** When `enforce == .clamp`, an out-of-range number is
///   silently pulled to the bound rather than 422'd. `clamp` says which ends, and
///   `clampNote` is the explanation to show — so "I typed 0.5 and got 1.0" is
///   never a mystery. When `enforce == .reject`, the server refuses instead, and
///   a client that helpfully clamps first would submit something other than what
///   was typed. Branch on `enforce`; never assume.
/// * **`lo`/`hi` are already operator-resolved.** The server reports
///   `min(declared_hi, operator_ceiling)`, so these are the bounds that will
///   actually be enforced. Don't re-derive them.
public struct Param: Sendable, Hashable, Codable, Identifiable {

    /// The closed set of value kinds (`KINDS` in param.py). Unknown kinds decode
    /// to `.unknown` rather than failing: a newer server adding a kind should
    /// degrade one control to a plain text field, not break the whole form.
    public enum Kind: Sendable, Hashable, Codable, RawRepresentable {
        case int, float, string, bool, csv, choice
        case date, url, urlList, regexList, secretRef, duration
        case unknown(String)

        public init?(rawValue: String) {
            switch rawValue {
            case "int": self = .int
            case "float": self = .float
            case "str": self = .string
            case "bool": self = .bool
            case "csv": self = .csv
            case "choice": self = .choice
            case "date": self = .date
            case "url": self = .url
            case "url_list": self = .urlList
            case "regex_list": self = .regexList
            case "secret_ref": self = .secretRef
            case "duration": self = .duration
            default: self = .unknown(rawValue)
            }
        }

        public var rawValue: String {
            switch self {
            case .int: return "int"
            case .float: return "float"
            case .string: return "str"
            case .bool: return "bool"
            case .csv: return "csv"
            case .choice: return "choice"
            case .date: return "date"
            case .url: return "url"
            case .urlList: return "url_list"
            case .regexList: return "regex_list"
            case .secretRef: return "secret_ref"
            case .duration: return "duration"
            case .unknown(let raw): return raw
            }
        }

        /// Whether values of this kind carry numeric bounds. Only these honour
        /// `lo`/`hi`/`clamp`.
        public var isNumeric: Bool {
            self == .int || self == .float
        }
    }

    /// Which UI control renders this param.
    ///
    /// The server's `EDITOR_FOR_KIND` supplies a default per kind, but the set is
    /// **open**: the registry passes a Module's declared `editor` string straight
    /// through, so a Pipeline can ask for a control no built-in setting uses. The
    /// cases below are the full vocabulary `DESIGN.md` §5.1 specifies a
    /// control for; anything else lands in `.unknown` and falls back to a text
    /// field, so a newer server adding a control degrades one row rather than
    /// breaking the form.
    public enum Editor: Sendable, Hashable, Codable, RawRepresentable {
        case number, textfield, checkbox, stringList, select
        case datepicker, url, regexList, secret, duration, textarea
        /// Menu of toggles, labelled from `enumTitles`.
        case multiselect
        /// Two-column table.
        case keyValue
        /// Text editor with parse-on-type and an inline error.
        case json
        /// Not rendered at all.
        case hidden
        case unknown(String)

        public init?(rawValue: String) {
            switch rawValue {
            case "number": self = .number
            case "textfield": self = .textfield
            case "checkbox": self = .checkbox
            case "stringList": self = .stringList
            case "select": self = .select
            case "datepicker": self = .datepicker
            case "url": self = .url
            case "regexList": self = .regexList
            case "secret": self = .secret
            case "duration": self = .duration
            case "textarea": self = .textarea
            case "multiselect": self = .multiselect
            case "keyValue": self = .keyValue
            case "json": self = .json
            case "hidden": self = .hidden
            default: self = .unknown(rawValue)
            }
        }

        public var rawValue: String {
            switch self {
            case .number: return "number"
            case .textfield: return "textfield"
            case .checkbox: return "checkbox"
            case .stringList: return "stringList"
            case .select: return "select"
            case .datepicker: return "datepicker"
            case .url: return "url"
            case .regexList: return "regexList"
            case .secret: return "secret"
            case .duration: return "duration"
            case .textarea: return "textarea"
            case .multiselect: return "multiselect"
            case .keyValue: return "keyValue"
            case .json: return "json"
            case .hidden: return "hidden"
            case .unknown(let raw): return raw
            }
        }

        /// A `hidden` param takes part in submission but draws nothing.
        public var isRendered: Bool { self != .hidden }
    }

    /// How the server handles an out-of-range value.
    public enum Enforce: String, Sendable, Hashable, Codable {
        /// Silently pulled to the bound. The windex default; right for settings
        /// and Pipeline Module configuration.
        case clamp
        /// Refused with a 422. Used where a value is an explicit instruction (a
        /// job argument) and running something else is worse than an error.
        case reject
    }

    /// Which ends of the range are silently enforced.
    public enum ClampEnds: String, Sendable, Hashable, Codable {
        case floor, ceiling, both

        public var clampsLow: Bool { self == .floor || self == .both }
        public var clampsHigh: Bool { self == .ceiling || self == .both }
    }

    /// Install-time vs run-time parameter (`"runtime"` for settings).
    public enum Stage: String, Sendable, Hashable, Codable {
        case install, runtime
    }

    /// A visibility condition on another field's value: `{"field": ..., "equals": ...}`
    /// or `{"field": ..., "in": [...]}`.
    public struct DependsOn: Sendable, Hashable, Codable {
        public let field: String
        public let equals: JSONValue?
        public let anyOf: [JSONValue]?

        enum CodingKeys: String, CodingKey {
            case field, equals
            case anyOf = "in"
        }

        public init(field: String, equals: JSONValue? = nil, anyOf: [JSONValue]? = nil) {
            self.field = field
            self.equals = equals
            self.anyOf = anyOf
        }

        /// Whether the dependent control should be shown, given the current value
        /// of `field` in the form. A condition naming neither `equals` nor `in` is
        /// treated as satisfied — an unparseable rule must not hide a control the
        /// operator then can't find.
        public func isSatisfied(by value: JSONValue?) -> Bool {
            if let equals { return value == equals }
            if let anyOf { return value.map(anyOf.contains) ?? false }
            return true
        }
    }

    // MARK: Identity

    /// The attribute/config name this param sets. Unique within its form.
    public let key: String
    public var id: String { key }
    public let kind: Kind

    // MARK: Presentation

    public let title: String
    public let description: String
    public let editor: Editor
    /// Groups fields under a heading. Empty means ungrouped.
    public let section: String?
    /// Rendered as a suffix ("s", "bytes").
    public let unit: String?
    /// Render inside a collapsed disclosure.
    public let advanced: Bool
    /// Write-only: never echoed on read, so an existing value shows as a
    /// placeholder rather than a populated field.
    public let secret: Bool
    /// Present ⇒ render disabled, with this as the explanation.
    public let lockedReason: String?

    // MARK: Values

    public let defaultValue: JSONValue?
    /// What a form STARTS with, distinct from `defaultValue` (what the server does
    /// when the key is absent). They differ where the default is unwieldy.
    public let prefill: JSONValue?
    public let required: Bool

    // MARK: Bounds — already tightened by the operator's ceiling/floor

    public let lo: Double?
    public let hi: Double?
    public let choices: [String]
    /// Human labels parallel to `choices`.
    public let enumTitles: [String]
    public let maxItems: Int?
    public let maxLength: Int?
    public let pattern: String?

    // MARK: Behaviour

    public let enforce: Enforce
    public let clamp: ClampEnds?
    public let clampNote: String?
    public let dependsOn: DependsOn?
    public let stage: Stage
    /// `secret_ref` only: the operator keys this param may name.
    public let allow: [String]

    // MARK: - Decoding

    private enum CodingKeys: String, CodingKey {
        case key, kind, lo, hi, choices, label, help
        case type, editor, title, description, required, advanced, secret, stage, enforce
        case defaultValue = "default"
        case prefill, section, unit, enumTitles, maxItems, maxLength, pattern
        case clamp, clampNote, lockedReason, dependsOn, allow
    }

    public init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        key = try c.decode(String.self, forKey: .key)
        kind = Kind(rawValue: try c.decode(String.self, forKey: .kind)) ?? .unknown("")

        // `describe()` emits both the legacy console keys (label/help) and the
        // generalized ones (title/description) with identical content. Prefer the
        // generalized pair and fall back, so this keeps decoding if the legacy
        // half is dropped once the console is deleted.
        title = try c.decodeIfPresent(String.self, forKey: .title)
            ?? c.decodeIfPresent(String.self, forKey: .label)
            ?? key
        description = try c.decodeIfPresent(String.self, forKey: .description)
            ?? c.decodeIfPresent(String.self, forKey: .help)
            ?? ""

        let rawEditor = try c.decodeIfPresent(String.self, forKey: .editor)
        editor = rawEditor.flatMap(Editor.init(rawValue:)) ?? Param.defaultEditor(for: kind)

        lo = try c.decodeIfPresent(Double.self, forKey: .lo)
        hi = try c.decodeIfPresent(Double.self, forKey: .hi)
        choices = try c.decodeIfPresent([String].self, forKey: .choices) ?? []
        enumTitles = try c.decodeIfPresent([String].self, forKey: .enumTitles) ?? []

        defaultValue = try c.decodeIfPresent(JSONValue.self, forKey: .defaultValue)
        prefill = try c.decodeIfPresent(JSONValue.self, forKey: .prefill)
        required = try c.decodeIfPresent(Bool.self, forKey: .required) ?? false
        advanced = try c.decodeIfPresent(Bool.self, forKey: .advanced) ?? false
        secret = try c.decodeIfPresent(Bool.self, forKey: .secret) ?? false

        section = try c.decodeIfPresent(String.self, forKey: .section)
        unit = try c.decodeIfPresent(String.self, forKey: .unit)
        maxItems = try c.decodeIfPresent(Int.self, forKey: .maxItems)
        maxLength = try c.decodeIfPresent(Int.self, forKey: .maxLength)
        pattern = try c.decodeIfPresent(String.self, forKey: .pattern)

        // An unrecognised enforce value must fail SAFE. `clamp` would let a client
        // quietly adjust a value the server then refuses; `reject` only costs an
        // error the operator can see.
        let rawEnforce = try c.decodeIfPresent(String.self, forKey: .enforce)
        enforce = rawEnforce.flatMap(Enforce.init(rawValue:)) ?? .reject
        clamp = try c.decodeIfPresent(ClampEnds.self, forKey: .clamp)
        clampNote = try c.decodeIfPresent(String.self, forKey: .clampNote)
        lockedReason = try c.decodeIfPresent(String.self, forKey: .lockedReason)
        dependsOn = try c.decodeIfPresent(DependsOn.self, forKey: .dependsOn)
        stage = try c.decodeIfPresent(Stage.self, forKey: .stage) ?? .runtime
        allow = try c.decodeIfPresent([String].self, forKey: .allow) ?? []
    }

    public func encode(to encoder: Encoder) throws {
        var c = encoder.container(keyedBy: CodingKeys.self)
        try c.encode(key, forKey: .key)
        try c.encode(kind.rawValue, forKey: .kind)
        try c.encode(title, forKey: .title)
        try c.encode(description, forKey: .description)
        try c.encode(editor.rawValue, forKey: .editor)
        try c.encode(required, forKey: .required)
        try c.encode(advanced, forKey: .advanced)
        try c.encode(secret, forKey: .secret)
        try c.encode(stage, forKey: .stage)
        try c.encode(enforce, forKey: .enforce)
        try c.encodeIfPresent(lo, forKey: .lo)
        try c.encodeIfPresent(hi, forKey: .hi)
        if !choices.isEmpty { try c.encode(choices, forKey: .choices) }
        if !enumTitles.isEmpty { try c.encode(enumTitles, forKey: .enumTitles) }
        try c.encodeIfPresent(defaultValue, forKey: .defaultValue)
        try c.encodeIfPresent(prefill, forKey: .prefill)
        try c.encodeIfPresent(section, forKey: .section)
        try c.encodeIfPresent(unit, forKey: .unit)
        try c.encodeIfPresent(maxItems, forKey: .maxItems)
        try c.encodeIfPresent(maxLength, forKey: .maxLength)
        try c.encodeIfPresent(pattern, forKey: .pattern)
        try c.encodeIfPresent(clamp, forKey: .clamp)
        try c.encodeIfPresent(clampNote, forKey: .clampNote)
        try c.encodeIfPresent(lockedReason, forKey: .lockedReason)
        try c.encodeIfPresent(dependsOn, forKey: .dependsOn)
        if !allow.isEmpty { try c.encode(allow, forKey: .allow) }
    }

    static func defaultEditor(for kind: Kind) -> Editor {
        switch kind {
        case .int, .float: return .number
        case .string: return .textfield
        case .bool: return .checkbox
        case .csv, .urlList: return .stringList
        case .choice: return .select
        case .date: return .datepicker
        case .url: return .url
        case .regexList: return .regexList
        case .secretRef: return .secret
        case .duration: return .duration
        case .unknown: return .textfield
        }
    }
}

// MARK: - Form behaviour

extension Param {
    /// Whether the control should be editable at all.
    public var isEditable: Bool { lockedReason == nil }

    /// The label to show for a choice, falling back to the raw value when the
    /// server didn't supply parallel titles.
    public func title(forChoice choice: String) -> String {
        guard let idx = choices.firstIndex(of: choice),
              idx < enumTitles.count else { return choice }
        return enumTitles[idx]
    }

    /// What an empty form field should start as: `prefill` when the server offered
    /// one (the default is unwieldy), otherwise `default`.
    public var initialValue: JSONValue? { prefill ?? defaultValue }

    /// Apply the server's clamp rule locally so the form can *preview* the
    /// adjustment before submitting.
    ///
    /// Returns `nil` when nothing would change. Only ever called for `.clamp`
    /// params — under `.reject` the server refuses out-of-range input, and
    /// adjusting it here would submit a value the operator didn't type.
    public func clamped(_ value: JSONValue) -> JSONValue? {
        guard enforce == .clamp, kind.isNumeric, let n = value.doubleValue else { return nil }
        var out = n
        if let lo, clamp?.clampsLow ?? true { out = max(out, lo) }
        if let hi, clamp?.clampsHigh ?? true { out = min(out, hi) }
        guard out != n else { return nil }
        return kind == .int ? .int(Int(out)) : .double(out)
    }

    /// Local pre-submit validation, mirroring `Param.coerce`'s rules so a form can
    /// show the error inline instead of round-tripping for a 422.
    ///
    /// This is a convenience, never the authority: `coerce()` on the server is the
    /// security boundary, and it runs regardless of what this returns.
    public func validate(_ value: JSONValue) -> String? {
        if let lockedReason { return "\(key): not editable (\(lockedReason))" }

        switch kind {
        case .bool:
            guard value.boolValue != nil else { return "\(key): expected true/false" }

        case .choice:
            guard let s = value.stringValue, choices.contains(s) else {
                return "\(key): must be one of \(choices.joined(separator: ", "))"
            }

        case .secretRef:
            guard let s = value.stringValue else { return "\(key): expected a string" }
            let name = s.trimmingCharacters(in: .whitespaces)
            if !allow.isEmpty && !allow.contains(name) {
                return "\(key): must be one of \(allow.joined(separator: ", "))"
            }

        case .string, .csv, .url, .date, .duration:
            guard let s = value.stringValue else { return "\(key): expected a string" }
            if let maxLength, s.count > maxLength {
                return "\(key): longer than \(maxLength) chars"
            }

        case .urlList, .regexList:
            guard let items = value.stringArrayValue else {
                return "\(key): expected a list of strings"
            }
            if let maxItems, items.count > maxItems {
                return "\(key): at most \(maxItems) items"
            }
            for item in items {
                if item.isEmpty { return "\(key): each item must be a non-empty string" }
                if let maxLength, item.count > maxLength {
                    return "\(key): item longer than \(maxLength) chars"
                }
            }
            if kind == .regexList {
                for pat in items where (try? NSRegularExpression(pattern: pat)) == nil {
                    return "\(key): invalid regex \(pat)"
                }
            }

        case .int, .float:
            // Python's coerce rejects bool for a numeric even though bool is an int
            // there; mirror that so the two agree on what's valid.
            if value.boolValue != nil { return "\(key): expected a number" }
            guard let n = value.doubleValue else { return "\(key): expected a number" }
            if kind == .int, case .double(let d) = value, d != d.rounded() {
                return "\(key): expected a number"
            }
            // Only `reject` produces an error; a `clamp` param out of range is
            // valid input that the server will quietly pull to the bound.
            if enforce == .reject {
                if (lo.map { n < $0 } ?? false) || (hi.map { n > $0 } ?? false) {
                    return "\(key) out of range [\(lo?.description ?? "-"), \(hi?.description ?? "-")]"
                }
            }

        case .unknown(let raw):
            return "\(key): unsupported kind '\(raw)'"
        }
        return nil
    }
}

File: src/windex/schema/param.py
"""One typed, bounded, self-describing parameter — the shape every form renders.

Three things in this codebase independently grew the same idea: `settings_schema.Field`
(editable settings), `api.jobs.Param` (job arguments), and `crawl.recipe`'s hand-written
clamp helpers (crawl limits). They agree on the important part — *a value is typed,
bounded, and clamped to the operator's ceiling rather than rejected* — and disagree on
everything else, so a client has to know three shapes to render three forms.

This is that idea once. `Param` is the superset; `describe()` is the JSON a client
builds a control from; `coerce()` is the validation, and it is a SECURITY BOUNDARY in
exactly the sense `settings_schema`'s docstring means: an allowlist, never a denylist,
because these values arrive over a LAN-exposed API.

Two rules carried over verbatim because they were learned the hard way:

* **Clamp, don't reject.** A caller may always ask to be slower or smaller; never
  faster or bigger than the operator's bound. Failing a whole form submit over a
  typo'd number is worse than silently honouring the ceiling — but the client is
  told, via `clamp`/`clampNote`, so "I typed 0.5 and got 1.0" is never a mystery.
* **`ceiling` / `floor` name an operator Settings key, not a number.** A recipe
  author writes `hi: 20000`; the operator's `crawl_max_pages_ceiling` is what
  actually caps it. The bound a caller sees is `min(hi, operator_ceiling)`. This is
  what lets an untrusted recipe declare generous bounds safely, and it is why
  ceiling keys are themselves absent from every editable allowlist.
"""

from __future__ import annotations

from dataclasses import dataclass, field

# The closed set of value kinds. Adding one means teaching `coerce`, `json_type`
# and every client's editor switch — deliberately a decision, not a default.
KINDS = (
    "int", "float", "str", "bool", "csv", "choice",
    "date", "url", "url_list", "regex_list", "secret_ref", "duration",
)

# Default UI control per kind. A Param may override via `editor` when the default
# reads wrong (a long `str` wanting a textarea, say).
EDITOR_FOR_KIND = {
    "int": "number", "float": "number", "str": "textfield", "bool": "checkbox",
    "csv": "stringList", "choice": "select", "date": "datepicker", "url": "url",
    "url_list": "stringList", "regex_list": "regexList", "secret_ref": "secret",
    "duration": "duration",
}

# The JSON type a value of each kind serializes as. Note `csv` is a STRING: it is
# stored as the raw comma-separated form that Settings' *_list() helpers parse, and
# rendering it as a list editor is a client-side affordance, not a storage change.
_JSON_TYPE = {
    "int": "integer", "float": "number", "str": "string", "bool": "boolean",
    "csv": "string", "choice": "string", "date": "string", "url": "string",
    "url_list": "array", "regex_list": "array", "secret_ref": "string",
    "duration": "string",
}

_NUMERIC = ("int", "float")
_STRINGY = ("str", "csv", "url", "date", "duration")
_LISTY = ("url_list", "regex_list")


@dataclass(frozen=True)
class Param:
    """One editable value. `key` is the attribute/config name it sets."""

    key: str
    kind: str
    lo: float | None = None
    hi: float | None = None
    choices: tuple[str, ...] = ()
    label: str = ""
    help: str = ""

    # --- value semantics ---
    default: object = None
    # `prefill` is what a form STARTS with, distinct from `default` (what the server
    # does when the key is absent). They differ where the default is unwieldy: the
    # DevDocs slug list has a large default nobody wants pasted into a text field.
    prefill: object = None
    required: bool = False

    # --- operator bounds: a Settings attribute name, resolved at coerce time ---
    ceiling: str | None = None
    floor: str | None = None

    # --- list/string bounds ---
    max_items: int | None = None
    max_len: int | None = None
    pattern: str = ""

    # --- form rendering ---
    editor: str = ""            # defaults from EDITOR_FOR_KIND
    section: str = ""           # groups fields under a heading
    unit: str = ""              # rendered as a suffix ("s", "bytes")
    advanced: bool = False      # render inside a collapsed disclosure
    secret: bool = False        # write-only; never echoed on read
    enum_titles: tuple[str, ...] = ()   # human labels parallel to `choices`

    # --- client behaviour hints ---
    locked_reason: str = ""     # present => render disabled with this explanation
    depends_on: dict | None = None      # {"field": "...", "equals"|"in": ...}
    clamp_note: str = ""        # shown when the value would be clamped

    # How an out-of-range value is handled. "clamp" is the windex default and the
    # right one for settings and recipe config. "reject" is for a value someone
    # typed as an explicit instruction — a job argument — where silently running
    # something other than what was asked is worse than an error. The client needs
    # to know which, or it will helpfully adjust a value the server then refuses.
    enforce: str = "clamp"      # clamp | reject

    # --- recipe-only ---
    stage: str = "runtime"      # install | runtime
    allow: tuple[str, ...] = field(default=())   # secret_ref: nameable operator keys

    def __post_init__(self) -> None:
        if self.kind not in KINDS:
            raise ValueError(f"{self.key}: unknown kind {self.kind!r}")

    # ------------------------------------------------------------------ describe

    @property
    def clamp(self) -> str:
        """Which ends of the range are silently enforced: floor | ceiling | both.

        Only meaningful for numerics, and it exists so a client can explain the
        adjustment before the server makes it.
        """
        if self.kind not in _NUMERIC or self.enforce != "clamp":
            return ""
        has_lo = self.lo is not None or self.floor is not None
        has_hi = self.hi is not None or self.ceiling is not None
        if has_lo and has_hi:
            return "both"
        if has_lo:
            return "floor"
        if has_hi:
            return "ceiling"
        return ""

    def to_spec(self) -> dict:
        """The STORAGE form — exactly the declaration, and nothing else.

        Distinct from `describe()` on purpose. `describe()` is for a client
        rendering a control: it resolves bounds against the operator's settings,
        camelCases for JS, and omits `ceiling`/`floor` because those name operator
        keys a client has no business seeing. None of that round-trips. A recipe's
        frozen spec must reconstruct the identical Param or a re-run is not a
        re-run, so this emits the declared values verbatim and omits defaults.
        """
        out = {"key": self.key, "kind": self.kind}
        for attr, default in (
            ("lo", None), ("hi", None), ("choices", ()), ("label", ""), ("help", ""),
            ("default", None), ("prefill", None), ("required", False),
            ("ceiling", None), ("floor", None), ("max_items", None),
            ("max_len", None), ("pattern", ""), ("editor", ""), ("section", ""),
            ("unit", ""), ("advanced", False), ("secret", False),
            ("enum_titles", ()), ("locked_reason", ""), ("depends_on", None),
            ("clamp_note", ""), ("stage", "runtime"), ("allow", ()),
            ("enforce", "clamp"),
        ):
            value = getattr(self, attr)
            if value != default:
                out[attr] = list(value) if isinstance(value, tuple) else value
        return out

    def describe(self, effective: object = None) -> dict:
        """The JSON a client renders a control from.

        `effective` is an optional Settings-like object; when given, `lo`/`hi` are
        reported as the OPERATOR-resolved bounds rather than the declared ones, so
        the form shows the range that will actually be enforced.
        """
        lo, hi = self.bounds(effective)
        out = {
            # --- legacy keys: settings_schema.Field.describe()'s exact shape.
            # api/static/components/sources.js reads these; keep them until the
            # console is deleted and the Swift client is the only reader.
            "key": self.key,
            "kind": self.kind,
            "lo": lo,
            "hi": hi,
            "choices": list(self.choices),
            "label": self.label or self.key,
            "help": self.help,
            # --- the generalized shape ---
            "type": _JSON_TYPE[self.kind],
            "editor": self.editor or EDITOR_FOR_KIND[self.kind],
            "title": self.label or self.key,
            "description": self.help,
            "required": self.required,
            "advanced": self.advanced,
            "secret": self.secret,
            "stage": self.stage,
            "enforce": self.enforce,
        }
        # Omit rather than emit nulls: an absent key means "not applicable here",
        # which is easier for a client to branch on than a null that might be a value.
        if self.default is not None:
            out["default"] = self.default
        if self.prefill is not None:
            out["prefill"] = self.prefill
        if self.section:
            out["section"] = self.section
        if self.unit:
            out["unit"] = self.unit
        if self.enum_titles:
            out["enumTitles"] = list(self.enum_titles)
        if self.max_items is not None:
            out["maxItems"] = self.max_items
        if self.max_len is not None:
            out["maxLength"] = self.max_len
        if self.pattern:
            out["pattern"] = self.pattern
        if self.clamp:
            out["clamp"] = self.clamp
        if self.clamp_note:
            out["clampNote"] = self.clamp_note
        if self.locked_reason:
            out["lockedReason"] = self.locked_reason
        if self.depends_on:
            out["dependsOn"] = dict(self.depends_on)
        if self.allow:
            out["allow"] = list(self.allow)
        return out

    # -------------------------------------------------------------------- bounds

    def bounds(self, effective: object = None) -> tuple[float | None, float | None]:
        """Declared bounds tightened by the operator's ceiling/floor keys.

        A missing Settings attribute is ignored rather than fatal: a recipe naming a
        ceiling key that a later windex removed should lose the extra tightening, not
        refuse to load. The declared `lo`/`hi` still bound it.
        """
        lo, hi = self.lo, self.hi
        if effective is not None:
            if self.floor:
                op = getattr(effective, self.floor, None)
                if op is not None:
                    lo = op if lo is None else max(lo, op)
            if self.ceiling:
                op = getattr(effective, self.ceiling, None)
                if op is not None:
                    hi = op if hi is None else min(hi, op)
        return lo, hi

    # -------------------------------------------------------------------- coerce

    def coerce(self, value, effective: object = None):
        """Validate + coerce + clamp one value. Raises ValueError (routes map to 422).

        Error message wording is load-bearing: `settings_schema`'s tests match on it,
        and it is what a form surfaces to the user.
        """
        key = self.key
        if self.locked_reason and value != self.default:
            # A lock means "you may not CHANGE this", not "this may not appear".
            # Restating the default is not a change — and it has to be allowed, or
            # a spec that materializes defaults (so a frozen run is complete and
            # self-describing) could never be parsed back in.
            raise ValueError(f"{key}: not editable ({self.locked_reason})")

        if self.kind == "bool":
            if not isinstance(value, bool):
                raise ValueError(f"{key}: expected true/false")
            return value

        if self.kind == "choice":
            if value not in self.choices:
                raise ValueError(f"{key}: must be one of {', '.join(self.choices)}")
            return value

        if self.kind == "secret_ref":
            if not isinstance(value, str):
                raise ValueError(f"{key}: expected a string")
            name = value.strip()
            # A secret_ref carries the NAME of an operator-provisioned key, never a
            # value, and only from a declared allowlist — so a recipe can neither
            # smuggle a credential in nor name one it was not offered.
            if self.allow and name not in self.allow:
                raise ValueError(f"{key}: must be one of {', '.join(self.allow)}")
            return name

        if self.kind in _STRINGY:
            if not isinstance(value, str):
                raise ValueError(f"{key}: expected a string")
            if self.kind == "csv":
                # Stored as the raw comma string Settings' *_list() helpers parse;
                # normalizing here is what keeps those helpers correct.
                out = ",".join(p.strip() for p in value.split(",") if p.strip())
            else:
                out = value.strip()
            if self.max_len is not None and len(out) > self.max_len:
                raise ValueError(f"{key}: longer than {self.max_len} chars")
            return out

        if self.kind in _LISTY:
            if isinstance(value, str):
                value = [value]
            if not isinstance(value, list):
                raise ValueError(f"{key}: expected a list of strings")
            if self.max_items is not None and len(value) > self.max_items:
                raise ValueError(f"{key}: at most {self.max_items} items")
            out_list = []
            for item in value:
                if not isinstance(item, str) or not item:
                    raise ValueError(f"{key}: each item must be a non-empty string")
                if self.max_len is not None and len(item) > self.max_len:
                    raise ValueError(
                        f"{key}: item longer than {self.max_len} chars")
                out_list.append(item)
            if self.kind == "regex_list":
                # Compile now so a bad pattern is a 422 at submit rather than an
                # exception inside a worker hours later.
                import re
                for pat in out_list:
                    try:
                        re.compile(pat)
                    except re.error as exc:
                        raise ValueError(f"{key}: invalid regex {pat!r} ({exc})")
            return out_list

        if self.kind in _NUMERIC:
            if isinstance(value, bool):      # bool is an int in Python; not here
                raise ValueError(f"{key}: expected a number")
            try:
                n = int(value) if self.kind == "int" else float(value)
            except (TypeError, ValueError):
                raise ValueError(f"{key}: expected a number")
            lo, hi = self.bounds(effective)
            if self.enforce == "reject":
                if (lo is not None and n < lo) or (hi is not None and n > hi):
                    raise ValueError(f"{key} out of range [{lo}, {hi}]")
                return n
            # Clamp rather than reject: a caller may ask to be gentler, and silently
            # honouring the bound beats failing a form submit over a typo.
            if lo is not None:
                n = max(n, int(lo) if self.kind == "int" else lo)
            if hi is not None:
                n = min(n, int(hi) if self.kind == "int" else hi)
            return n

        raise ValueError(f"{key}: unsupported kind {self.kind!r}")
